docs/apis/subsystems/external/writing-a-service.md

```diff
@@ -316,10 +316,10 @@ Because some web service protocols are strict about the number and types of argu
 ```php
 public static function get_biscuit_parameters() {
     return new external_function_parameters([
-        'chocolatechips' => new external_value(PARAM_BOOL, PARAM_REQUIRED),
-        'glutenfree' => new external_value(PARAM_BOOL, PARAM_DEFAULT, false),
+        'chocolatechips' => new external_value(PARAM_BOOL, 'if biscuit contains chocolate chips', PARAM_REQUIRED),
+        'glutenfree' => new external_value(PARAM_BOOL, 'if biscuit is glutenfree', PARAM_DEFAULT, false),
```

Comment:
The other thing you can demonstrate is named args:

```suggestion
        'glutenfree' => new external_value(
            type: PARAM_BOOL,
            default: PARAM_DEFAULT,
            allownull: false,
        ),
```

Reply:
Also, this one is missing _two_ args, the sig is:

```php
    public function __construct(
        $type,
        $desc = '',
        $required = VALUE_REQUIRED,
        $default = null,
        $allownull = NULL_ALLOWED
    );
```

PARAM_DEFAULT is the 4th param, not the 3rd.

Reply:
I used the named args, but your suggestion for `default: PARAM_DEFAULT` looks wrong to me. Shouldn't it be it `required: VALUE_DEFAULT, default: false`?

Reply:
Here's the origial code:
```
        'chocolatechips' => new external_value(PARAM_BOOL, PARAM_REQUIRED),
```

And your version:
```
        'chocolatechips' => new external_value(PARAM_BOOL, 'if biscuit contains chocolate chips', PARAM_REQUIRED),
```

So there is no `PARAM_REQUIRED` - it's `VALUE_REQUIRED`

And for `glutenfree`:
```
        'glutenfree' => new external_value(PARAM_BOOL, 'if biscuit is glutenfree', PARAM_DEFAULT, false),
```

There is no `PARAM_DEFAULT`, it's `VALUE_DEFAULT`

So it's anyone's guess what the original code was trying to suggest. I suspect you're right, but that the contants are just incorrect and shoudl be VALUE_ not PARAM_.

Reply:
Yes, true. As I said my whole changes now should fix this confusion in the whole example.